doc/man/recollq.1
.\" $Id: recollq.1,v 1.1 2007-11-13 10:07:35 dockes Exp $ (C) 2005 J.F.Dockes\$
.TH RECOLLQ 1 "13 November 2007"
.SH NAME
recollq \- command line / standard output Recoll query command.
.SH SYNOPSIS
.B recollq
[
.B \-c
<configdir>
]
[
.B \-o
|
.B \-f
|
.B \-a
]
[
.B \-b
]
[
.B \-d
]
[
.B \-A
]
[
.B \-p
]
[
.B \-e
]
[
.B \-m
]
[
.B \-n
<[first-]cnt>
]
[
.B \-Q
]
[
.B \-s
<stemming language>
]
[
.B \-S
<fldname>
]
[
.B \-D
]
[
.B \-i
<additional index directory>
]
[
.B \-F
<quoted space separated field name list>
]
<query string>

.B recollq \-P

.SH DESCRIPTION
The
.B recollq
command will execute the Recoll query specified on the command line and
print the results to the standard output. It is primarily designed for
diagnostics, or piping the data to some other program. The basic format and
its variations can be useful for command line querying. The \-F option
should exclusively be used for using the output data in another program, as
it is the only one for which output is guaranteed to be fully parseable.
.PP
The 
.B \-c 
option specifies the configuration directory name, overriding the
default or $RECOLL_CONFDIR.
.PP
The query string is built by concatenating all arguments found at the end
of the command line (after the options). It will be interpreted by default
as a query language string. Quoting should be used as needed to escape
characters that might be interpreted by the shell (ie: wildcards).
.B \-a 
is specified, the query string will be interpreted as an
.I all words
simple search query. If 
.B \-o 
is specified, the query string will be interpreted as an
.I any word
simple search query. If 
.B \-f
is specified, the query string will be interpreted as a
.I file name
simple search query. 
.PP
.B \-b
(basic) can be specified to only print the result urls in the output
stream.
.PP
If 
.B \-d
is set, the text for the result files contents will be dumped to stdout.
.PP
If 
.B \-m
is set, the whole metadata array will be dumped for each document.
.PP
If 
.B \-A
is set, the document abstracts will be printed. With an additional
.B \-p
option, snippets with page numbers (when available) will be shown instead.
.PP
.B \-S
<fieldname>
sorts the results according to the specified field. Use 
.B \-D 
for descending order.
.PP
.B \-n
<cnt>
can be used to set the maximum number of results that should be
printed. The default is 2000. Use a value of 0 for no limit.
.PP
.B \-s
<language>
selects the word stemming language. The value should match an existing
stemming database (as set in the configuration or added with recollindex \-s).
.PP
.B \-i
<extra dbdir>
adds the specified Xapian index to the set used for the query. Can be
specified multiple times.
.PP 
.B \-F
<space separated field list> (one argument, e.g. "author title")
should be used for piping the data to another program. After 2 initial
lines showing the actual query and the estimated result counts, it will
print one line for each result document. Each line will have 
exactly the fields requested on the command line. Fields are encoded in
base64 and separated by one space character. Empty fields are indicated by
consecutive space characters. There is one additional space character at
the end of each line.
.PP
.B recollq \-P
(Period) will print the minimum and maximum modification years for
documents in the index.

.SH SEE ALSO
.PP 
recollindex(1) recollq(1) recoll.conf(5) 


ChangeLog
2010-02-02 15:33 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (d11da0283f03 [tip])

	* src/common/textsplit.cpp, src/common/textsplit.h,
	src/query/plaintorich.cpp, src/query/recollq.cpp,
	src/query/wasatorcl.cpp, src/rcldb/rcldb.cpp,
	src/rcldb/searchdata.cpp, src/rcldb/stoplist.cpp,
	src/rcldb/stoplist.h:
	cosmetics: use derived class for actual splitter instead of callback

2010-02-02 10:24 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (a8caf709bcd3)

	* src/qt4gui/rclmain.ui, src/qtgui/rclmain.ui,
	src/qtgui/rclmain_w.cpp:
	QT GUI: define accelerators for res list page movements

2010-02-02 08:20 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (ec31e285a553)

	* src/qtgui/reslist.cpp, src/query/reslistpager.h:
	Qt GUI: ensure that new page size is taken into account ASAP (no
	need for restarting app)

2010-02-01 17:51 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (db953bb94c7f)

	* src/qt4gui/rclmain.ui, src/qtgui/preview_w.cpp,
	src/qtgui/rclmain.ui, src/qtgui/rclmain_w.cpp,
	src/qtgui/rclmain_w.h:
	QT GUI: add fullscreen mode

2010-02-01 10:31 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (1eda55ae3be9)

	* src/mk/manifest.txt: new file.
	* src/excludefile, src/makesrcdist.sh, src/mk/manifest.txt:
	Making a source dist: check the new list against old reference +
	other checks

2010-01-31 19:53 +0100  Israel G. Lugo  <Israel G. Lugo>  (74d4e25d43c2)

	* src/recollinstall.in:
	Install recollq and its manpage when in cmdline mode. Don't install
	the recoll.1 manpage when in cmdline mode.

2010-01-31 19:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (c88b0ef40512)

	* src/common/autoconfig.h.in:
	use 3-arg version of ac_define as the 1-arg one is being obsoleted

2010-01-31 19:45 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (1960435ccb68)

	* src/configure.ac:
	Dispense with the x11-monitoring when neither fam nor inotify are
	configured

2010-01-31 19:35 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (08e6abfc5fdf)

	* src/configure.ac:
	use 3-arg version of ac_define as the 1-arg one is being obsoleted

2010-01-31 19:34 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (c0add9dd8ad4)

	* website/download.html:
	none

2010-01-30 17:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (5ed138ff2230)

	* src/qtgui/spell_w.cpp, src/qtgui/spell_w.h:
	QT GUI: fix small problems in newly native qt4 term expander

2010-01-30 17:31 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (6ecf959a8e01)

	* src/qt4gui/spell.ui: new file.
	* src/qt4gui/recollmain.ui: deleted file.
	* .hgignore, src/qt4gui/recollmain.ui, src/qt4gui/spell.ui,
	src/qt4gui/uifrom3, src/qtgui/spell_w.cpp, src/qtgui/spell_w.h:
	QT GUI: converted the qt4 term expander dialog to native qt4

2010-01-30 14:09 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (df8a91aaff88)

	* src/qt4gui/rclmain.ui: new file.
	* .hgignore, src/qt4gui/rclmain.ui, src/qt4gui/uifrom3,
	src/qtgui/confgui/confguiindex.h, src/qtgui/rclmain_w.cpp,
	src/qtgui/rclmain_w.h:
	Qt GUI: made the Qt4 main window native Qt4 (no more use of
	Q3MainWindow)

2010-01-30 08:23 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (ed18703563b7)

	* .hgignore, src/ChangeLog, website/BUGS.html, website/devel.html,
	website/doc.html, website/download.html, website/index.html.en:
	none

2010-01-30 08:23 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (22044a3b2e2c)

	* src/qtgui/rclmain_w.h:
	Qt 4.6.1 Uic bug: change qt version test from == to <= as bug still
	here in 4.6.2

2010-01-30 08:21 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (b1cb8c664953)

	* src/common/autoconfig.h.in, src/configure, src/configure.ac,
	src/index/recollindex.cpp, src/utils/x11mon.cpp:
	Renamed WITHOUT_X11 to DISABLE_X11MON for clarification

2010-01-30 08:18 +0100  Israel G. Lugo  <Israel G. Lugo>  (be03b72e1258)

	* src/configure.ac:
	Rename option without-gui to disable-qtgui. New option disable-
	x11mon. Separate control of creation of the gui and X11 session
	monitoring.

2010-01-29 19:00 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (d95c21312a15)

	* src/kde/kioslave/recoll/CMakeLists.txt:
	KIO slave: fixed CMakeList to configure Recoll with --enable-pic

2010-01-29 17:22 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (266941720a99)

	* src/python/README.txt: new file.
	* src/configure, src/configure.ac, src/doc/user/usermanual.sgml,
	src/lib/Makefile, src/lib/mkMake, src/mk/commondefs,
	src/mk/localdefs.in, src/php/00README.txt, src/python/README.txt,
	src/python/recoll/setup.py:
	Implemented configure --enable-pic flag to build the main lib with
	position-independent objects. This avoids having to edit localdefs
	by hand to build the new php extension, and voids the need for the
	Python module to recompile Recoll source files.

2010-01-29 15:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (69c42078b8d3)

	* src/php/00README.txt: new file.
	* src/php/00README.txt, src/php/recoll/make.sh,
	src/php/recoll/recoll.cpp:
	PHP extension by Wenqiang Song : make ready for external use.
	- added minimal doc
	- fixed build script to work around php/libtool issue
	- have the module default to Query Language (instead of AND)

2010-01-28 18:22 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (45e7ec5e16c5)

	* .hgignore, website/usermanual/README-dir.txt: new file.
	* packaging/debian/changelog, packaging/debian/compat,
	packaging/debian/control, packaging/debian/copyright,
	packaging/debian/docs, packaging/debian/menu,
	packaging/debian/rules, packaging/debian/watch: deleted file.
	* .hgignore, packaging/debian/changelog, packaging/debian/compat,
	packaging/debian/control, packaging/debian/copyright,
	packaging/debian/docs, packaging/debian/menu,
	packaging/debian/rules, packaging/debian/watch, src/makesrcdist.sh,
	website/usermanual/README-dir.txt:
	svn->mercurial modifications

2010-01-28 16:13 +0000  convert-repo  <convert-repo>  (e85c82d42126)

	* .hgtags: new file.
	* .hgtags:
	update tags

2010-01-26 13:23 +0000  dockes  <dockes>  (c0cb63a2702a)


query/recollq.cpp
/* Copyright (C) 2006 J.F.Dockes
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the
 *   Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
// Takes a query and run it, no gui, results to stdout
#include "autoconfig.h"

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <limits.h>

#include <iostream>
#include <sstream>
#include <list>
#include <string>

#include "rcldb.h"
#include "rclquery.h"
#include "rclconfig.h"
#include "pathut.h"
#include "rclinit.h"
#include "log.h"
#include "wasatorcl.h"
#include "internfile.h"
#include "wipedir.h"
#include "transcode.h"
#include "textsplit.h"
#include "smallut.h"
#include "chrono.h"
#include "base64.h"

using namespace std;

bool dump_contents(RclConfig *rclconfig, Rcl::Doc& idoc)
{
    FileInterner interner(idoc, rclconfig, FileInterner::FIF_forPreview);
    Rcl::Doc fdoc;
    string ipath = idoc.ipath;
    if (interner.internfile(fdoc, ipath)) {
        cout << fdoc.text << endl;
    } else {
        cout << "Cant turn to text:" << idoc.url << " | " << idoc.ipath << endl;
    }
    return true;
}

string make_abstract(Rcl::Doc& doc, Rcl::Query& query, bool asSnippets,
                     int snipcount)
{
    string abstract;
    if (asSnippets) {
        std::vector<Rcl::Snippet> snippets;
        std::ostringstream str;
        if (query.makeDocAbstract(doc, snippets, snipcount, -1, true)) {
            for (const auto& snippet : snippets) {
                str << snippet.page << " : " << snippet.snippet << endl;
            }
        }
        abstract = str.str();
    } else {
        query.makeDocAbstract(doc, abstract);
        abstract += "\n";
    }
    return abstract;
}

void output_fields(vector<string> fields, Rcl::Doc& doc,
                   Rcl::Query& query, Rcl::Db&, bool printnames,
                   bool asSnippets, int snipcnt)
{
    if (fields.empty()) {
        map<string,string>::const_iterator it;
        for (const auto& entry : doc.meta) {
            fields.push_back(entry.first);
        }
    }
    for (vector<string>::const_iterator it = fields.begin();
         it != fields.end(); it++) {
        string out;
        if (!it->compare("abstract")) {
            base64_encode(make_abstract(doc, query, asSnippets, snipcnt), out);
        } else if (!it->compare("xdocid")) {
            char cdocid[30];
            sprintf(cdocid, "%lu", (unsigned long)doc.xdocid);
            base64_encode(cdocid, out);
        } else {
            base64_encode(doc.meta[*it], out);
        }
        // Before printnames existed, recollq printed a single blank for empty
        // fields. This is a problem when printing names and using strtok, but
        // have to keep the old behaviour when printnames is not set.
        if (!(out.empty() && printnames)) {
            if (printnames)
                cout << *it << " ";
            cout << out << " ";
        }
    }
    cout << endl;
}

static char *thisprog;
static char usage [] =
" -P: Show the date span for all the documents present in the index.\n"
" [-o|-a|-f] [-q] <query string>\n"
" Runs a recoll query and displays result lines. \n"
"  Default: will interpret the argument(s) as a xesam query string.\n"
"  Query elements: \n"
"   * Implicit AND, exclusion, field spec:  t1 -t2 title:t3\n"
"   * OR has priority: t1 OR t2 t3 OR t4 means (t1 OR t2) AND (t3 OR t4)\n"
"   * Phrase: \"t1 t2\" (needs additional quoting on cmd line)\n"
" -o Emulate the GUI simple search in ANY TERM mode.\n"
" -a Emulate the GUI simple search in ALL TERMS mode.\n"
" -f Emulate the GUI simple search in filename mode.\n"
" -q is just ignored (compatibility with the recoll GUI command line).\n"
"Common options:\n"
" -c <configdir> : specify config directory, overriding $RECOLL_CONFDIR.\n"
" -C : collapse duplicates\n"            
" -d also dump file contents.\n"
" -n [first-]<cnt> define the result slice. The default value for [first]\n"
"    is 0. Without the option, the default max count is 2000.\n"
"    Use n=0 for no limit.\n"
" -b : basic. Just output urls, no mime types or titles.\n"
" -Q : no result lines, just the processed query and result count.\n"
" -m : dump the whole document meta[] array for each result.\n"
" -A : output the document abstracts.\n"
"    -p <cnt> : show <cnt> snippets, with page numbers instead of the concatenated abstract.\n"
" -S fld : sort by field <fld>.\n"
"   -D : sort descending.\n"
" -s stemlang : set stemming language to use (must exist in index...).\n"
"    Use -s \"\" to turn off stem expansion.\n"
" -T <synonyms file>: use the parameter (Thesaurus) for word expansion.\n"
" -i <dbdir> : additional index, several can be given.\n"
" -e use url encoding (%xx) for urls.\n"
" -E use exact result count instead of lower bound estimate.\n"
" -F <field name list> : output exactly these fields for each result.\n"
"    The field values are encoded in base64, output in one line and \n"
"    separated by one space character. This is the recommended format \n"
"    for use by other programs. Use a normal query with option -m to \n"
"    see the field names. Use -F '' to output all fields, but you probably\n"
"    also want option -N in this case.\n"
"  -N : with -F, print the (plain text) field names before the field values.\n"
;

static void
Usage(void)
{
    cerr << thisprog <<  ": usage:" << endl << usage;
    exit(1);
}

// BEWARE COMPATIBILITY WITH recoll OPTIONS letters
static int     op_flags;

#define OPT_A     0x1
// GUI: -a same
#define OPT_a     0x2
#define OPT_b     0x4
#define OPT_C     0x8
// GUI: -c same
#define OPT_c     0x10 
#define OPT_D     0x20 
#define OPT_d     0x40 
#define OPT_e     0x80 
#define OPT_F     0x100
// GUI: -f same
#define OPT_f     0x200
// GUI uses -h for help. us: usage
#define OPT_i     0x400
// GUI uses -L to set language of messages
// GUI: -l same
#define OPT_l     0x800
#define OPT_m     0x1000
#define OPT_N     0x2000
#define OPT_n     0x4000
// GUI: -o same
#define OPT_o     0x8000
#define OPT_p     0x10000
#define OPT_P     0x20000
#define OPT_Q     0x40000
// GUI: -q same
#define OPT_q     0x80000
#define OPT_S     0x100000
#define OPT_s     0x2000000
#define OPT_T     0x4000000
// GUI: -t use command line, us: ignored
#define OPT_t     0x800000
// GUI uses -v : show version. Us: usage
// GUI uses -w : open minimized
#define OPT_E     0x1000000

int recollq(RclConfig **cfp, int argc, char **argv)
{
    string a_config;
    string sortfield;
    string stemlang("english");
    list<string> extra_dbs;
    string sf;
    vector<string> fields;
    string syngroupsfn;
    
    int firstres = 0;
    int maxcount = 2000;
    int snipcnt = -1;
    
    thisprog = argv[0];
    argc--; argv++;

    while (argc > 0 && **argv == '-') {
        (*argv)++;
        if (!(**argv))
            /* Cas du "adb - core" */
            Usage();
        while (**argv)
            switch (*(*argv)++) {
            case '-': 
                // -- : end of options
                if (*(*argv) != 0)
                    Usage();
                goto endopts;
            case 'A':   op_flags |= OPT_A; break;
            case 'a':   op_flags |= OPT_a; break;
            case 'b':   op_flags |= OPT_b; break;
            case 'C':   op_flags |= OPT_C; break;
            case 'c':   op_flags |= OPT_c; if (argc < 2)  Usage();
                a_config = *(++argv);
                argc--; goto b1;
            case 'd':   op_flags |= OPT_d; break;
            case 'D':   op_flags |= OPT_D; break;
            case 'E':   op_flags |= OPT_E; break;
            case 'e':   op_flags |= OPT_e; break;
            case 'f':   op_flags |= OPT_f; break;
            case 'F':   op_flags |= OPT_F; if (argc < 2)  Usage();
                sf = *(++argv);
                argc--; goto b1;
            case 'i':   op_flags |= OPT_i; if (argc < 2)  Usage();
                extra_dbs.push_back(*(++argv));
                argc--; goto b1;
            case 'l':   op_flags |= OPT_l; break;
            case 'm':   op_flags |= OPT_m; break;
            case 'N':   op_flags |= OPT_N; break;
            case 'n':   op_flags |= OPT_n; if (argc < 2)  Usage();
            {
                string rescnt = *(++argv);
                string::size_type dash = rescnt.find("-");
                if (dash != string::npos) {
                    firstres = atoi(rescnt.substr(0, dash).c_str());
                    if (dash < rescnt.size()-1) {
                        maxcount = atoi(rescnt.substr(dash+1).c_str());
                    }
                } else {
                    maxcount = atoi(rescnt.c_str());
                }
                if (maxcount <= 0) maxcount = INT_MAX;
            }
            argc--; goto b1;
            case 'o':   op_flags |= OPT_o; break;
            case 'P':   op_flags |= OPT_P; break;
            case 'p':
            {
                op_flags |= OPT_p;
                if (argc < 2)
                    Usage();
                const char *cp = *(++argv);
                char *cpe;
                snipcnt = strtol(cp, &cpe, 10);
                if (*cpe != 0)
                    Usage();
                argc--; goto b1;
            }
            case 'q':   op_flags |= OPT_q; break;
            case 'Q':   op_flags |= OPT_Q; break;
            case 'S':   op_flags |= OPT_S; if (argc < 2)  Usage();
                sortfield = *(++argv);
                argc--; goto b1;
            case 's':   op_flags |= OPT_s; if (argc < 2)  Usage();
                stemlang = *(++argv);
                argc--; goto b1;
            case 't':   op_flags |= OPT_t; break;
            case 'T':   op_flags |= OPT_T; if (argc < 2)  Usage();
                syngroupsfn = *(++argv);
                argc--; goto b1;
            default: Usage();   break;
            }
    b1: argc--; argv++;
    }
endopts:

    string reason;
    *cfp = recollinit(0, 0, 0, reason, &a_config);
    RclConfig *rclconfig = *cfp;
    if (!rclconfig || !rclconfig->ok()) {
        fprintf(stderr, "Recoll init failed: %s\n", reason.c_str());
        exit(1);
    }

    if (argc < 1 && !(op_flags & OPT_P)) {
        Usage();
    }
    if (op_flags & OPT_F) {
        if (op_flags & (OPT_b|OPT_d|OPT_b|OPT_Q|OPT_m|OPT_A))
            Usage();
        stringToStrings(sf, fields);
    }
    Rcl::Db rcldb(rclconfig);
    if (!extra_dbs.empty()) {
        for (list<string>::iterator it = extra_dbs.begin();
             it != extra_dbs.end(); it++) {
            if (!rcldb.addQueryDb(*it)) {
                cerr << "Can't add index: " << *it << endl;
                exit(1);
            }
        }
    }
    if (!syngroupsfn.empty()) {
        if (!rcldb.setSynGroupsFile(syngroupsfn)) {
            cerr << "Can't use synonyms file: " << syngroupsfn << endl;
            exit(1);
        }
    }
    
    if (!rcldb.open(Rcl::Db::DbRO)) {
        cerr << "Cant open database in " << rclconfig->getDbDir() << 
            " reason: " << rcldb.getReason() << endl;
        exit(1);
    }

    if (op_flags & OPT_P) {
        int minyear, maxyear;
        if (!rcldb.maxYearSpan(&minyear, &maxyear)) {
            cerr << "maxYearSpan failed: " << rcldb.getReason() << endl;
            exit(1);
        } else {
            cout << "Min year " << minyear << " Max year " << maxyear << endl;
            exit(0);
        }
    }

    if (argc < 1) {
        Usage();
    }
    string qs = *argv++;argc--;
    while (argc > 0) {
        qs += string(" ") + *argv++;argc--;
    }

    {
        string uq;
        string charset = rclconfig->getDefCharset(true);
        int ercnt;
        if (!transcode(qs, uq, charset, "UTF-8", &ercnt)) {
            fprintf(stderr, "Can't convert command line args to utf-8\n");
            exit(1);
        } else if (ercnt) {
            fprintf(stderr, "%d errors while converting arguments from %s "
                    "to utf-8\n", ercnt, charset.c_str());
        }
        qs = uq;
    }

    std::shared_ptr<Rcl::SearchData> sd;

    if (op_flags & (OPT_a|OPT_o|OPT_f)) {
        sd = std::make_shared<Rcl::SearchData>(Rcl::SCLT_OR, stemlang);
        Rcl::SearchDataClause *clp = 0;
        if (op_flags & OPT_f) {
            clp = new Rcl::SearchDataClauseFilename(qs);
        } else {
            clp = new Rcl::SearchDataClauseSimple((op_flags & OPT_o)?
                                                  Rcl::SCLT_OR : Rcl::SCLT_AND, 
                                                  qs);
        }
        if (sd)
            sd->addClause(clp);
    } else {
        sd = wasaStringToRcl(rclconfig, stemlang, qs, reason);
    }

    if (!sd) {
        cerr << "Query string interpretation failed: " << reason << endl;
        return 1;
    }

    std::shared_ptr<Rcl::SearchData> rq(sd);
    Rcl::Query query(&rcldb);
    if (op_flags & OPT_C) {
        query.setCollapseDuplicates(true);
    }
    if (op_flags & OPT_S) {
        query.setSortBy(sortfield, (op_flags & OPT_D) ? false : true);
    }
    Chrono chron;
    if (!query.setQuery(rq)) {
        cerr << "Query setup failed: " << query.getReason() << endl;
        return(1);
    }
    int cnt;
    if (op_flags & OPT_E) {
        cnt = query.getResCnt(-1, true);
    } else {
        cnt = query.getResCnt();
    }
    if (!(op_flags & OPT_b)) {
        cout << "Recoll query: " << rq->getDescription() << endl;
        if (firstres == 0) {
            if (cnt <= maxcount)
                cout << cnt << " results" << endl;
            else
                cout << cnt << " results (printing  " << maxcount << " max):" 
                     << endl;
        } else {
            cout << "Printing at most " << cnt - (firstres+maxcount) <<
                " results from first " << firstres << endl;
        }
    }
    if (op_flags & OPT_Q)
        cout << "Query setup took " << chron.millis() << " mS" << endl;

    if (op_flags & OPT_Q)
        return(0);

    for (int i = firstres; i < firstres + maxcount; i++) {
        Rcl::Doc doc;
        if (!query.getDoc(i, doc))
            break;

        if (op_flags & OPT_F) {
            output_fields(fields, doc, query, rcldb,
                          op_flags & OPT_N, op_flags & OPT_p, snipcnt);
            continue;
        }

        if (op_flags & OPT_e) 
            doc.url = url_encode(doc.url);

        if (op_flags & OPT_b) {
            cout << doc.url << endl;
        } else {
            string titleorfn = doc.meta[Rcl::Doc::keytt];
            if (titleorfn.empty())
                titleorfn = doc.meta[Rcl::Doc::keyfn];
            if (titleorfn.empty()) {
                string url;
                printableUrl(rclconfig->getDefCharset(), doc.url, url);
                titleorfn = path_getsimple(url);
            }

            char cpc[20];
            sprintf(cpc, "%d", doc.pc);
            cout 
                << doc.mimetype << "\t"
                << "[" << doc.url << "]" << "\t" 
                << "[" << titleorfn << "]" << "\t"
                << doc.fbytes << "\tbytes" << "\t"
                <<  endl;
            if (op_flags & OPT_m) {
                for (const auto& ent : doc.meta) {
                    cout << ent.first << " = " << ent.second << endl;
                }
            }
            if (op_flags & OPT_A) {
                bool asSnippets = (op_flags & OPT_p) != 0;
                string abstract = make_abstract(doc, query, asSnippets, snipcnt);
                string marker = asSnippets ? "SNIPPETS" : "ABSTRACT";
                if (!abstract.empty()) {
                    cout << marker << endl;
                    cout << abstract;
                    cout << string("/") + marker << endl;
                }
            }
        }
        if (op_flags & OPT_d) {
            dump_contents(rclconfig, doc);
        }   
    }

    return 0;
}
